frr-metrics WIP

oribon opened the pull request:
TODO once merged:
* Add build to dev-env and speaker image
* Integrate with the manifests (drop prometheus port anno, add sidecar ...)
* change e2e tests?

fedepaol commented on frr-metrics/collector/bgp.go:
```diff
@@ -0,0 +1,140 @@
+	bgpNamespace = "metallb"
+	bgpSubsystem = "bgp"
+
+	sessionUpDesc = prometheus.NewDesc(
```
This seems to be lower level than using CounterVec / GaugeVec as the current implementation does.
Would it make sense to have the content of https://github.com/metallb/metallb/blob/2683861f566bf01fb3111cf37d6241f87bfca1dd/internal/bgp/stats.go#L5 exported as a package and consumable from here?

The benefit would be having one single place describing the metrics.

fedepaol commented on frr-metrics/collector/bgp.go:
```diff
@@ -0,0 +1,140 @@
+		return
+	}
+
+	neighbors, err := parseNeighbours(res)
```
If I get this right, we are fetching the metrics in the very same moment prometheus asks us for them.
What I normally saw (but this doesn't necessarily mean that is the right way to do it), is that we put values in the GaugeVec (or other "bucket") and then prometheus scrapes the content of it.

Here, we are not event driven because we are pulling info from frr, but having a periodic goroutine that does the scraping and fills the buckets would make the trick. Prometheus scrape interval is normally long enough (like 30 seconds), so a goroutine that updates every 10 seconds would suffice I think.

oribon committed:
Move BGP FRR parsing to internal

Moving this content as a different package under internal/
and exporting it allows us to reuse it from different places
such as for collecting FRR Prometheus metrics.

Signed-off-by: Ori Braunshtein <obraunsh@redhat.com>

oribon committed:
FRR: Add Prometheus exporter

* In order to keep feature-parity with the current
BGP implementation we need to collect and export
the same Prometheus metrics.

* With FRR, we can't use the same approach to
collect the metrics as we currently do by producing the metrics ourselves
(https://prometheus.io/docs/introduction/glossary/#direct-instrumentation).
Rather we need to "collect" them from FRR, which is best done
using a Prometheus exporter and writing a collector for the BGP daemon
(https://prometheus.io/docs/instrumenting/writing_exporters/).
Each time Prometheus scrapes the /metrics endpoint the metrics are
gathered by the collector from FRR using vtysh.
The collector needs to satisfy the Prometheus.Collector interface.

* Adding new collectors to the exporter per FRR daemon
is possible in the future.

* The exporter will be built and shipped with the
speaker, and would be deployed as a seperate sidecar
container that has access to the FRR sockets like the reloader.

Signed-off-by: Ori Braunshtein <obraunsh@redhat.com>
